### PR TITLE
[mlir][IR] Add `Block::isReachable` helper function

### PR DESCRIPTION
**Pull request by matthias-springer**
Add a new helper function `isReachable` to `Block`. This function traverses all successors of a block to determine if another block is reachable from the current block.

This functionality has been reimplemented in multiple places in MLIR. Possibly additional copies in downstream projects. Therefore, moving it to a common place.

### CODE REVIEW

**Review by River707 — CHANGES_REQUESTED** [review]
Why don't the various transforms just use DominanceInfo? Answering these queries is what it's intended for.

**Comment by River707** [comment]
Coming back to this PR after reading https://github.com/llvm/llvm-project/pull/114940 (which uses this function). I don't think we should add this kind of thing to Block, I'd rather push people to use DominanceInfo (which does this more optimally). Things like Dialect Conversion (which make a muck of blocks and all kinds of other things) are generally very unique compared to the majority of users. For API's on the core data structures, the APIs should (generally) be pushing for the best way to do something.

**Comment by matthias-springer** [comment]
> Why don't the various transforms just use DominanceInfo? Answering these queries is what it's intended for.

Block reachability and block dominance are two different things. Just because there is a path from `block A -> block B` does not necessarily mean that `block A dominates block B`. I'm not sure how "block reachability" could be implemented with `DominanceInfo`.

That being said, in the dialect conversion I actually need dominance, so this helper function is actually not useful there...

File: mlir/lib/IR/Block.cpp [hunk 1]
```diff
@@ -349,6 +352,23 @@ SuccessorRange::SuccessorRange(Operation *term) : SuccessorRange() {
...
 }
 
+bool Block::isReachable(Block *other, ArrayRef<Block *> except) {
+  assert(getParent() == other->getParent() && "expected same region");
+  SmallVector<Block *> worklist(succ_begin(), succ_end());
+  SmallPtrSet<Block *, 16> visited;
+  while (!worklist.empty()) {
+    Block *next = worklist.pop_back_val();
+    if (llvm::is_contained(except, next))
+      continue;
```
Comment (joker-eph) [inline mlir/lib/IR/Block.cpp:362]: This is a linear search, should this be the last check of the three?

File: mlir/lib/IR/Block.cpp [hunk 2]
```diff
@@ -349,6 +352,23 @@ SuccessorRange::SuccessorRange(Operation *term) : SuccessorRange() {
     base = term->getBlockOperands().data();
 }
 
+bool Block::isReachable(Block *other, ArrayRef<Block *> except) {
+  assert(getParent() == other->getParent() && "expected same region");
+  SmallVector<Block *> worklist(succ_begin(), succ_end());
+  SmallPtrSet<Block *, 16> visited;
```
Comment (joker-eph) [inline mlir/lib/IR/Block.cpp:358]: What about having the "except" parameters be a `SmallPtrSetImpl<Block *>` passed by copy (user should move) and use it to store the visited elements.
That way the user can still pre-populate it with blocks to skip, but we have one less check and one less data-structure around.

**Review by River707 — APPROVED** [review]
LG!

**Pull request merged**